### DISCUSSION TITLE
Modify Keycloak source code without recompiling the whole project

### DISCUSSION THREAD

**Discussion by Bredstone** [keycloak/keycloak#14101] [Q&A]
Hello! :)

I'm trying to add some post quantum algorithms to Keycloak via custom SPI's and, to do so, I need to modify some core classes (mostly **Provider** classes) from Keycloak with my code so the system recognizes the new algorithms.

Unfortunately, recompile Keycloak is a toilsome job *(that takes a lot of time)*, therefore modifying core files would make impracticable to run tests and stuff, while developing.

I've been studying Keycloak source files and noticed that the providers I need to modify *(apparently)* don't implement **ProviderFactory** interfaces, so *(I think)* it's not possible to modify these files with a SPI approach.

My questions are: Is it possible to modify such classes without recompiling entire Keycloak? Can I use custom SPI's to fix these problems? Should I try another solution? *(I'm kinda out of ideas)*

By the way, one of the classes I need to modify is **BCPemUtilsProvider** and I would like to do something like this:
``` java
    @Override
    protected String encode(Object obj) {
        if (obj == null) {
            return null;
        }

        try {
            StringWriter writer = new StringWriter();
            JcaPEMWriter pemWriter = new JcaPEMWriter(writer);

            /** 
             * MY CODE GOES HERE...
             */

            pemWriter.writeObject(obj);
            pemWriter.flush();
            pemWriter.close();
            String s = writer.toString();
            return removeBeginEnd(s);
        } catch (Exception e) {
            throw new PemException(e);
        }
    }
```

**Comment by stelewis-redhat** [comment]
What you could do is convert those places where certain "Providers" aren't pluggable into official providers with a lookup. That would take some extra effort, but the Provider API is relatively straightforward, and then you could iterate more quickly in a non-core package, just to prove your ideas out.

Take my suggestion with a grain of salt. I'm not a maintainer, and I have no idea if maintainers would approve of my suggestion, but technically, I don't see why it couldn't be done.
